### PR TITLE
Update S3 extention codes

### PR DESCRIPTION
**Pull request by adam8157**
I did some impovements of s3 after last s3 pull request. This one is to update its code and move development to GPDB source tree .

> compatible with s3.amazonaws.com and s3-us-east-1.amazonaws.com regions
> able to cancel query while downloading
> catch any exceptions thrown by C++ functions
> set min and max values of threadnum and chunksize
> 
> and will move development to GPDB source tree after this.

Cc: @lij55 @yydzero @50wu

### CODE REVIEW

**Comment by lij55** [comment]
Could you attach test result and coverage of this latest enhancement on master?

**Comment by adam8157** [comment]
Attached, looks good.

[unittests_coverage_local.txt](https://github.com/greenplum-db/gpdb/files/152105/unittests_coverage_local.txt)

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp [hunk 1]
```diff
@@ -9,9 +9,14 @@
...
 
+// invoked by s3_import(), need to be exception safe
 S3ExtBase *CreateExtWrapper(const char *url) {
-    S3ExtBase *ret = new S3Reader(url);
-    return ret;
+    try {
+        S3ExtBase *ret = new S3Reader(url);
+        return ret;
+    } catch (...) {
+        return NULL;
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: Shall we log the exception we have for troubleshooting purpose?

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp [hunk 2]
```diff
@@ -41,55 +46,62 @@ S3Reader::S3Reader(string url) : S3ExtBase(url) {
...
+                    return false;
+                }
             }
+            break;
         }
-        break;
+        S3INFO("Got %d files to download", this->keylist->contents.size());
+        this->getNextDownloader();
+    } catch (...) {
+        return false;
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: Shall we log exception we have?

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp [hunk 3]
```diff
@@ -199,10 +226,18 @@ bool S3ExtBase::ValidateURL() {
...
         return false;
+    } else if (ibegin == string::npos) {
+        this->region = "external-1";
+    } else {
+        this->region = url.substr(ibegin + 1, iend - ibegin - 1);
+    }
+
+    if (this->region.compare("us-east-1") == 0) {
+        this->region = "external-1";
     }
```
Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: What to do to support new AWS region? Could we make it configurable in file or in static array?

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp (hunk 1 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On Mon, Feb 29, 2016 at 09:12:25PM -0800, Yandong Yao wrote:
> In gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp:
> 
> >  S3ExtBase *CreateExtWrapper(const char *url) {
> > -    S3ExtBase *ret = new S3Reader(url);
> > -    return ret;
> > +    try {
> > +        S3ExtBase *ret = new S3Reader(url);
> > +        return ret;
> > +    } catch (...) {
> > +        return NULL;
> 
> Shall we log the exception we have for troubleshooting purpose?

We could not, there is no class which covers all exceptions.

-- 
Adam Lee

Comment (yydzero) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: could we log something before return, so that we knows there are exception
occurs, even though we don't know what is the exact exception?

On Tue, Mar 1, 2016 at 2:14 PM, Adam Lee <notifications@github.com> wrote:

> In gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp
> <https://github.com/greenplum-db/gpdb/pull/445#discussion_r54526748>:
>
> >  S3ExtBase *CreateExtWrapper(const char *url) {
> > -    S3ExtBase *ret = new S3Reader(url);
> > -    return ret;
> > +    try {
> > +        S3ExtBase *ret = new S3Reader(url);
> > +        return ret;
> > +    } catch (...) {
> > +        return NULL;
>
> On Mon, Feb 29, 2016 at 09:12:25PM -0800, Yandong Yao wrote: In
> gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp: > S3ExtBase
> *CreateExtWrapper(const char *url) { > - S3ExtBase *ret = new
> S3Reader(url); > - return ret; > + try { > + S3ExtBase *ret = new
> S3Reader(url); > + return ret; > + } catch (...) { > + return NULL; Shall
> we log the exception we have for troubleshooting purpose?
> We could not, there is no class which covers all exceptions.
> … <#1864817026_>
> -- Adam Lee
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/445/files#r54526748>.
>



-- 
Best Regards,
Yandong

Comment (adam8157) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On Mon, Feb 29, 2016 at 10:15:48PM -0800, Yandong Yao wrote:
> could we log something before return, so that we knows there are exception
> occurs, even though we don't know what is the exact exception?

Make sense, will find a proper C thread-safe logging function to do
that.

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp (hunk 3 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On Mon, Feb 29, 2016 at 09:15:31PM -0800, Yandong Yao wrote:
> What to do to support new AWS region? Could we make it configurable in file or
> in static array?

To support signature v4. Wrong regions would return connection error,
but static array is a good idea, will consider about it.

Comment (hlinnaka) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On 01/03/16 06:24, Adam Lee wrote:
>>           return false;
>> +    } else if (ibegin == string::npos) {
>> +        this->region = "external-1";
>> +    } else {
>> +        this->region = url.substr(ibegin + 1, iend - ibegin - 1);
>> +    }
>> +
>> +    if (this->region.compare("us-east-1") == 0) {
>> +        this->region = "external-1";
>>       }
>
> On Mon, Feb 29, 2016 at 09:15:31PM -0800, Yandong Yao wrote:
>> What to do to support new AWS region? Could we make it configurable in file or
>> in static array?
>
> To support signature v4. Wrong regions would return connection error,
> but static array is a good idea, will consider about it.

The code doesn't need the name of the region for anything, only the 
endpoint URL. It would be simplest to let the user specify the endpoint 
URL directly, and use it as is.

- Heikki

Comment (danielgustafsson) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: The region should be completely configurable, users in Europe will not want to use US regions etc.

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp [hunk 4]
```diff
@@ -137,50 +148,66 @@ string S3Reader::getKeyURL(const string key) {
...
+
     return true;
 }
 
 bool S3ExtBase::ValidateURL() {
+    // TODO http://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
+    // s3.eu-central-1.amazonaws.com -> s3-eu-central-1.amazonaws.com
+    // s3.ap-northeast-2.amazonaws.com -> s3-ap-northeast-2.amazonaws.com
+
```
Comment (danielgustafsson) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: This comment needs to be expanded upon to make sense.

File: gpAux/extensions/gps3ext/src/S3Downloader.cpp [hunk 5]
```diff
@@ -184,6 +185,11 @@ void *DownloadThreadfunc(void *data) {
     uint64_t filled_size = 0;
     S3INFO("Downloading thread starts");
     do {
+        if (QueryCancelPending) {
+            S3INFO("Downloading thread is interrupted by GPDB");
+            return NULL;
+        }
+
```
Comment (danielgustafsson) [inline gpAux/extensions/gps3ext/src/S3Downloader.cpp]: Cancelling of queries seem like a new feature rather than a bugfix or am I missing something? If this is a new feature it should be in a separate commit.

**Comment by danielgustafsson** [comment]
Please split these fixes/features into separate commits rather than one large commit.

File: gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp (hunk 3 above)
Comment (adam8157) [inline gpAux/extensions/gps3ext/src/S3ExtWrapper.cpp]: On Tue, Mar 01, 2016 at 10:24:33AM -0800, Daniel Gustafsson wrote:
> The region should be completely configurable, users in Europe will not want to
> use US regions etc.

Sure, users could specify other regions by url, codes here are just
processing some domains.

**Comment by adam8157** [comment]
Thanks to everyone, will take the suggestions and submit another pull request.

@hlinnaka, I will review your codes later and take all the good ideas :)
@danielgustafsson, I was developing on anoter submodule repo which has different hierarchy, planing to move to GPDB source tree after this. Will try to separate them, thanks.
